### PR TITLE
feat: support backend tls settings with envoyproxy

### PR DESCRIPTION
**Pull request by alexwo**
**What this PR does / why we need it**:
This change allows to enhance the security and adaptability of connections to backend services. By allowing administrators to specify TLS settings, such as which cryptographic ciphers and TLS protocol versions to use, it allows to better align their security posture with industry best practices and regulatory requirements.

**Approach**:
- Enable specification of default values through a TLS settings section in the EnvoyProxy backend.
- Use these settings as the default TLS configurations whenever a BackendTLS policy is specified for a backend, applying the EnvoyProxy defaults to affected backends.

Once the BackendTLS policy or an alternative method for providing these settings is implemented/accepted in upstream, opt for more specific settings than those provided by envoyproxy backendtls settings.

**Which issue(s) this PR fixes**:
https://github.com/envoyproxy/gateway/issues/2901

### CODE REVIEW

**Comment by guydc** [comment]
cc @zhaohuabing who was also looking into backend MTLS in  #2984

File: internal/xds/translator/utils.go [hunk 1]
```diff
@@ -161,3 +161,17 @@ func createExtServiceXDSCluster(rd *ir.RouteDestination, tCtx *types.ResourceVer
 	}
 	return nil
 }
+
+func ParseTLSProtocol(tlsVersion ir.TLSVersion) (tlsv3.TlsParameters_TlsProtocol, error) {
```
Comment (guydc) [inline internal/xds/translator/utils.go:165]: can we reuse the util that we have in CTP for this mapping?

File: internal/gatewayapi/testdata/envoyproxy-tls-settings.in.yaml [hunk 2]
```diff
@@ -0,0 +1,133 @@
+gateways:
```
Comment (guydc) [inline internal/gatewayapi/testdata/envoyproxy-tls-settings.in.yaml:1]: Can we also add MTLS + TLS Settings tests for ext-auth and ext-proc, demonstrating how this applied to these sorts of backends?

**Review by guydc — COMMENTED** [review]
Thanks for this!
Some comments on reuse of existing types.

**Comment by arkodg** [comment]
thanks for working on this @alexwo !
this PR is doing two things
1. added support for specifying `clientCert` which doesnt exist in upstream today
2. added support for specifying tls params such as `cipher_suites`

both 1. and 2. are added into the global `EnvoyProxy` API
should 2. live inside BTP, so it can be modified on a per route (per upstream) basis ?

File: api/v1alpha1/envoyproxy_types.go [hunk 3]
```diff
@@ -116,6 +117,34 @@ type EnvoyProxySpec struct {
...
+// BackendTLSConfig describes the BackendTLS configuration for Envoy Proxy.
+type BackendTLSConfig struct {
+	// ClientCertificateRef defines the reference to a Kubernetes Secret that contains
+	// the client certificate and private key for Envoy to use when connecting to
+	// backend services and external services, such as ExtAuth, ALS, OpenTelemetry, etc.
+	// +optional
+	ClientCertificateRef *gwapiv1.SecretObjectReference `json:"clientCertificateRef,omitempty"`
+	// Ciphers defines only support the specified cipher list when negotiating BackendTLS 1.0-1.2 (this setting has no effect when negotiating BackendTLS 1.3).
+	// +optional
+	Ciphers []string `json:"ciphers,omitempty"`
```
[the remark this answers is not in the document]
Comment (alexwo) [reply api/v1alpha1/envoyproxy_types.go:135]: We can re-use it, however, it also hot ClientValidation and ALPNProtocol, this are less interesting in this context. 
Would it be better to re-structure TLSSettings and include the common only?

File: internal/gatewayapi/backendtlspolicy.go [hunk 4]
```diff
@@ -78,7 +78,26 @@ func (t *Translator) processBackendTLSPolicy(
 	}
 
 	status.SetAcceptedForPolicyAncestors(&policy.Status, ancestorRefs, t.GatewayControllerName)
-
+	// apply defaults as per envoyproxy
+	if resources.EnvoyProxy != nil {
```
[the remark this answers is not in the document]
Comment (alexwo) [reply internal/gatewayapi/backendtlspolicy.go:82]: Right, if support for additional resources is implemented in upstream, we can move / extend this logic to other areas as well. Do you see an advantage in making this change now? If we don’t currently need to apply it o the gateway level for example, perhaps we can delay this refactor till gateway add support or we see a use case. wdyt?

**Comment by alexwo** [comment]
> thanks for working on this @alexwo ! this PR is doing two things
> 
> 1. added support for specifying `clientCert` which doesnt exist in upstream today
> 2. added support for specifying tls params such as `cipher_suites`
> 
> both 1. and 2. are added into the global `EnvoyProxy` API should 2. live inside BTP, so it can be modified on a per route (per upstream) basis ?

Hi @arkodg ,

Probably this make sense, 

> thanks for working on this @alexwo ! this PR is doing two things
> 
> 1. added support for specifying `clientCert` which doesnt exist in upstream today
> 2. added support for specifying tls params such as `cipher_suites`
> 
> both 1. and 2. are added into the global `EnvoyProxy` API should 2. live inside BTP, so it can be modified on a per route (per upstream) basis ?
Hi @arkodg ,
This will feature better flexibility this way, perhaps it can make more sense for (2) like settings, @guydc WDYT?

File: internal/xds/translator/utils.go (hunk 1 above)
Comment (alexwo) [reply internal/xds/translator/utils.go:165]: yes, i think we should be able i will check it.

**Comment by guydc** [comment]
> It can allow for better flexibility, perhaps it can make more sense for (2) settings, @guydc WDYT?

The major downside I see in using BTP is that it can only attach to xRoute backends. So, if we need to apply TLS settings for other backends (extproc, extauth, OTEL, ALS ... ), we will need to add TLS setting in the scope of of #3069. 

I'm fine with both options, as long as we can support TLS settings for both xRoute and non-xRoute backends. @arkodg - WDYT?

**Comment by arkodg** [comment]
> > It can allow for better flexibility, perhaps it can make more sense for (2) settings, @guydc WDYT?
> 
> The major downside I see in using BTP is that it can only attach to xRoute backends. So, if we need to apply TLS settings for other backends (extproc, extauth, OTEL, ALS ... ), we will need to add TLS setting in the scope of of #3069.
> 
> I'm fine with both options, as long as we can support TLS settings for both xRoute and non-xRoute backends. @arkodg - WDYT?

yeah for non xRoute we'll need to embed it into our `backendRef`

by moving 2. into BTP, we add the ability to fine tune settings, but now `tls` settings are spread out in 3 places, which can get confusing
* `clientCertificateRef` in `EnvoyProxy.Spec.BackendTLS`
* `tlsParams` in `BackendTrafficPolicy.Spec.TLS`
* `caCertificateRefs` in `BackendTLSPolicy.Spec.Validation`

**Comment by alexwo** [comment]
> * BackendTLSPolicy

Does it make sense to move clientCertificateRef, to BTP as well?  this way we can have things less spread.

**Comment by guydc** [comment]
> If we go with the BTP direction, does it make sense to move clientCertificateRef, to BTP as well? This way we can have things less spread.

I think so. It's possible that different backends will require a different client cert. It's not uncommon for organizations to use private PKI(s) for client certs, and so backends may have different trust stores and CAs. If we go with EnvoyProxy-level config, we basically assume that all backends can be configured to trust the EG client certificate.

File: api/v1alpha1/envoyproxy_types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (arkodg) [reply api/v1alpha1/envoyproxy_types.go:135]: we can create a struct that represents the `TLSParams` and embed that struct in both place and handle the `ClientCertificateRef` and `ClientValidation` fields seperately

**Comment by arkodg** [comment]
@guydc @alexwo as platform administrators, would you say that the feature is necessary, because you are trying to enforce specific tls parameters (ciphers, tls versions). If this is true, then the argument of putting it in `EnvoyProxy.Spec.BackendTLS` can be made for 2. since for this use case, we'd want the platform admin to enforce these settings w/o allowing the app dev (route level BTP owner) to override it.

**Comment by guydc** [comment]
> as platform administrators, would you say that the feature is necessary, because you are trying to enforce specific tls parameters (ciphers, tls versions)

Yes, our use case is related to defining system-wide upstream TLS settings, and so EnvoyProxy makes a lot of sense to me.

**Comment by arkodg** [comment]
> > as platform administrators, would you say that the feature is necessary, because you are trying to enforce specific tls parameters (ciphers, tls versions)
> 
> Yes, our use case is related to defining system-wide upstream TLS settings, and so EnvoyProxy makes a lot of sense to me.

cool, i'm a +1 for this as well

File: api/v1alpha1/envoyproxy_types.go [hunk 5]
```diff
@@ -116,6 +117,20 @@ type EnvoyProxySpec struct {
...
+}
+
+// BackendTLSConfig describes the BackendTLS configuration for Envoy Proxy.
+type BackendTLSConfig struct {
+	// ClientCertificateRef defines the reference to a Kubernetes Secret that contains
+	// the client certificate and private key for Envoy to use when connecting to
+	// backend services and external services, such as ExtAuth, ALS, OpenTelemetry, etc.
+	// +optional
+	ClientCertificateRef *gwapiv1.SecretObjectReference `json:"clientCertificateRef,omitempty"`
+	TLSSettings          `json:",inline"`
```
[the remark this answers is not in the document]
Comment (alexwo) [reply api/v1alpha1/envoyproxy_types.go:133]: In this context, it make sense as everything is under BackendTLS, probably its less clear if we do something like: backendTLS.tls.

**Review by guydc — APPROVED** [review]
LGTM, Thanks!

File: internal/xds/translator/translator.go [hunk 6]
```diff
@@ -768,10 +768,29 @@ func buildXdsUpstreamTLSCASecret(tlsConfig *ir.TLSUpstreamConfig) *tlsv3.Secret
 func buildXdsUpstreamTLSSocketWthCert(tlsConfig *ir.TLSUpstreamConfig) (*corev3.TransportSocket, error) {
 
 	var tlsCtx *tlsv3.UpstreamTlsContext
+	var tlsParams *tlsv3.TlsParameters
```
Comment (arkodg) [inline internal/xds/translator/translator.go:771]: can we reuse https://github.com/envoyproxy/gateway/blob/27a770f4db8ff275591934f4bc3080ba0d1aaddc/internal/xds/translator/listener.go#L572 ?

Comment (alexwo) [reply internal/xds/translator/translator.go]: /retest

**Review by guydc — APPROVED** [review]

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Comment by sherwinwangs** [comment]
> clientCertificateRef

that would be a better design, how someone will be contribute to this feature

**Pull request merged**